homeassistant/config_entries.py

```diff
@@ -1694,3 +1699,22 @@ async def support_remove_from_device(hass: HomeAssistant, domain: str) -> bool:
     integration = await loader.async_get_integration(hass, domain)
     component = integration.get_component()
     return hasattr(component, "async_remove_config_entry_device")
+
+
+async def _async_get_application_credential_id(
```

Comment:
I'm assuming we don't want this depending directly on application credentials, hence why this works this way. Please let me know if there is a better approach for structuring the dependencies, as this was just my best guess to get discussion started.

Reply:
I don't think the config entries should know about application credentials. We may need an abstract interface to accomplish the remove context details. I don't have a specific idea just yet.

Reply:
I've proposed a hook that integrations can register to add additional data on removal.

This can run any time during remove (not just post remove) so if there is a simpler semantic we can use that is useful in more context we can do that too (e.g. a variation on async_remove_entry that includes a remove value)